Introduce `ext.dota2` - Dota 2 Game Coordinator extension

Aluerie opened the pull request:
## 🐸 Summary

This PR introduces `ext.dota2` extension. 

## 🖍️ TODO

### 🐍 Migrate ValvePython/dota2 implemented methods.

* 👨‍💼**Player**
  - [X] request_profile
  - [X] ~request_gc_profile~ (disabled by valve)
  - [X] request_profile_card
  - [ ] request_player_stats
  - [ ] request_player_info (disabled by valve?)
  - [ ] request_conduct_scorecard (behavior score system was revamped though)
  - [ ] request_hero_standings
* ⚽**Match**
  - [x] request_matchmaking_stats
  - [x] request_match_details
  - [ ] request_matches
  - [x] request_matches_minimal
  - [X] request_top_source_tv_games
  - [x] request_player_match_history
* 🥳**Party** 
  Sorry, I'm not interested in implementing this atm.
* 🦞 **Lobby** 
  Same^
* 🗣️**Chat**
  Same^
* 🤝**Shared Objects**
  - [ ] Maybe?
* 🛠️**Utils**
  - [x] replay_url
  - [x] replay_url_from_match
  - [x] metadata_url
  - [x] metadata_url_from_match

## 🔚 Checklist

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [X] If code changes were made then they have been tested.
  - [X] I have updated the documentation to reflect the changes. (a bit?)
- [ ] This PR fixes an issue.
- [X] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

Aluerie committed:
🪧Compile protobufs into `.py` mirrors

Done with `protoc -I . --python_betterproto_out=protos dota_gcmessages_client_watch.proto` using 1.2.5 betterproto and manual edits

Aluerie committed:
🐸Initial implementation for Dota2 GC

only introduces one method `fetch_top_source_tv_games`

Aluerie committed:
📃Initial docs for ext.dota2

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie committed:
✒️Stealth edit to `fetch_top_source_tv_games` docstring

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Gobot1234 commented on docs/ext/dota2/index.md:
```diff
@@ -0,0 +1,11 @@
+(steam-ext-dota2)=
+
+# `steam.ext.dota2` - A Team Fortress 2 Game Coordinator client
```
```suggestion
# `steam.ext.dota2` - A Dota 2 Game Coordinator client
```

Gobot1234 commented on steam/ext/dota2/client.py:
```diff
@@ -0,0 +1,127 @@
+
+        Parameters
+        ----------
+        search_key : :class: `str`, optional
+            Unknown purpose.
+        league_id : :class: `int`, optional
+            `league_id` for the professional tournament.
+        hero_id : :class: `int`, optional
+            `hero_id` to filter results by, just like in-client Watch Tab feature.
+        start_game : :class: `Literal[0, 10, 20, 30, 40, 50, 60, 70, 80, 90]`, optional, by default 0
+            This argument controls how many responses the game coordinator should return.
+            For example, `start_game=0` returns a list with 1 Response,
+            `start_game=90` returns a list of 10 Responses.
+            Game coordinator fills each response in consequent manner with games that satisfy keyword arguments
+            or with currently live highest average MMR games in case no other argument was given.
+        game_list_index : :class: `int`, optional
+            Only get responses matching `game_list_index`. Responses from Game Coordinator change
+            from time to time. The `game_list_index` indicates that those responses belong to the same chunk.
+        lobby_ids : :class: `list[int]`, optional
+            Lobby ids to query against.
```
```suggestion
        league_id
            The league ID for the professional tournament.
        hero_id
            `hero_id` to filter results by which can be found in the client Watch Tab feature.
        start_game
            This argument controls how many responses the game coordinator should return.
            For example, `start_game=0` returns a list with 1 Response,
            `start_game=90` returns a list of 10 Responses.
            The GC fills each response in consequent manner with games that satisfy keyword arguments
            or with currently live highest average MMR games in case no other argument was given.
        game_list_index
            Only get responses matching `game_list_index`. Responses from GC change
            from time to time. The `game_list_index` indicates that those responses belong to the same chunk.
        lobby_ids
            The lobby IDs of the games to fetch.
```
The functionality for start_game doesn't really make much sense, would it be possible to have this just be a limit argument and it does whatever maths it needs to for you?

Aluerie replied:
actually, it kinda works like that, I guess i will rewrite this sentence, 

i'm not sure if I want to rename the param name - as I said, I kinda thought it would be nice to have it 1x1 matching valve docs. But yeah to something like `total_games` would make much more sense for what this thing is for.

Gobot1234 commented on steam/ext/dota2/client.py:
```diff
@@ -0,0 +1,127 @@
+    _APP: Final = DOTA2
+    _state: GCState  # type: ignore  # PEP 705
+
+    async def fetch_top_source_tv_games(
```
```suggestion
    async def top_source_tv_games(
```
This can drop fetch_ because there isn't a get equivalent

Gobot1234 commented on steam/ext/dota2/client.py:
```diff
@@ -0,0 +1,127 @@
+
+        Returns
+        -------
+        list[dota_gcmessages_client_watch.CMsgGCToClientFindTopSourceTVGamesResponse]
+            The list of responses from Dota 2 Coordinator.
+            Those are grouped into grouped into chunks of 10 games.
```
```suggestion
        The responses from the GC. They are grouped into grouped into chunks of 10. games.
```
Why are they chunked into 10, and can we not just return as a model that isn't directly from the GC?

Aluerie replied:
I guess my wording sucks. It is just how this protobuf message works (by Valve). We send one message to GC and if for example
* `start_game=90` then game coordinator will send 10 messages back.
* if `start_game = 50` - it will send 6 messages back (Each message contains group-info about 10 games and some info about each game too. )
* if we requested `start_game=90, hero_id=1` and GC found 22 games then first 2 reponses will have 10 games, 3rd will have 2 and the rest is empty (but still have group-info).  

So I just add those responses in a list of responses to return

Idk, I kinda didn't want to go away from GC models or protobufs. I would rather want to have a clear way to use protobufs without rethinking Valve's structure or naming, like I basically send kwargs to the proto and get responses back.

Gobot1234 replied:
Am I correct in saying its start_game//10 + 1 sort of? I think if the messages are empty and don't contain any new data we shouldn't be returning them.

I really don't agree with a lot of Valve's namings of things which is a lot of the reason I created this library, very few methods (if any) return the protobufs directly because it just makes things harder to use especially because they don't have any sort of deprecation period or policy like I have now.

Gobot1234 commented on steam/ext/dota2/client.py:
```diff
@@ -0,0 +1,127 @@
+
+        await self._state.ws.send_gc_message(CMsgClientToGCFindTopSourceTVGames(**kwargs))
+
+        try:
+            async with timeout(30.0):
+                return await asyncio.gather(*futures)
+        except asyncio.TimeoutError:
+            raise asyncio.TimeoutError(
+                "Request time-outed. The reason might be inappropriate combination of keyword arguments, "
+                + "inappropriate argument values or simply Dota 2 Game Coordinator being down."
+            )
```
```suggestion
        async with timeout(30.0):
            return await asyncio.gather(*futures)
```
I don't think we should be adding the note here, if they want to know why it failed they should look at the docs

Gobot1234 commented on steam/ext/dota2/client.py:
```diff
@@ -0,0 +1,127 @@
+
+            return predicate
+
+        futures = [
+            self._state.ws.gc_wait_for(
+                CMsgGCToClientFindTopSourceTVGamesResponse,
+                check=start_game_check(game_counter),
+            )
+            for game_counter in list(range(0, start_game + 1, 10))
+        ]
```
```suggestion
        futures = [
            self._state.ws.gc_wait_for(
                CMsgGCToClientFindTopSourceTVGamesResponse,
                check=partial(lambda start_game, msg: msg.start_game == start_game, start_game),
            )
            for start_game in range(0, start_game + 1, 10)
		]
```
This should probably use functools.partial(predicate, start_game) and use the same variable name as the for loop

Aluerie replied:
my god, i knew i m forgetting something simple

Gobot1234 reviewed (changes requested):
Thank you so much for this!

![image](https://github.com/Gobot1234/steam.py/assets/50501825/a0021b4a-478a-46dd-a127-74e2714f0dda)
These don't need to be included, they are for steam non-sense.


Also if you wouldn't mind please could you clean up the enum names to remove underscores and normalise them to be PascalCase.

Finally could you add a very barebones test to the tests directory just to ensure that the client is importable

Aluerie committed:
Apply suggestions from code review by Gobot

Co-authored-by: James Hilton-Balfe <gobot1234yt@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie committed:
Apply suggestions from code review

Co-authored-by: James Hilton-Balfe <gobot1234yt@gmail.com>

Aluerie committed:
⛔Remove Steam NonSense

Aluerie committed:
🔨Fix bad enum namings

Aluerie committed:
🪓 Total Overhaul of the PR

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie commented:
Thanks for the feedback!

Last time I had similar conversation (with Riot Games API wrapper developers) they actually had the opposite opinion - that it's better for both library devs and end-user to just follow "bad naming/practices/models" from the original devs instead of rethinking/over-cooking. The proto-models are here and kinda working - why not just give easy access to them, they aren't so bad, honestly. 

Maybe affected by those, I just thought 
* to compile all protos 
* to list all known emsg-response pairs 
* to have easy kwargs->proto interface and let the user do whatever with returned responses.
* only then implement helper methods which would modelize the responses.

But I definitely see your points/philosophy. Sorry that I was too linear. Let's modelize this stuff right away then. I did a total overhaul of the PR (maybe I should resubmit it for cleaner commit history?). This is what I changed:

* On practice, this top source proto message has very little keyword combinations that work and make sense. Thus, I reimplemented this into 3 separate methods. 
* I also implemented some models to wrap responses and its fields.
* deleted all unused enums and proto messages -> naming and steam nonsense problems are gone.
* intial test file. 
* moved EMsg to `enums.py` in root of the extension (kinda like yours).
* renamed some files to feel more pythonic and less valve-gibbersih prefixes (kinda like yours).

Aluerie committed:
🖋️Small doc edit

Gobot1234 commented on steam/ext/dota2/protobufs/shared_enums.py:
I would move these to an enums.py in dota2 to follow the other packages as these aren't used anywhere in the protos

Gobot1234 commented on steam/ext/dota2/protobufs/lobby.py:
Same comment for shared_enums

Gobot1234 commented on steam/ext/dota2/protobufs/shared_enums.py:
```diff
@@ -0,0 +1,79 @@
+
+# fmt: off
+class DOTAGameMode(IntEnum):
+    NONE              = 0
+    AP                = 1
+    CM                = 2
+    RD                = 3
+    SD                = 4
+    AR                = 5
+    INTRO             = 6
+    HW                = 7
+    REVERSE_CM        = 8
+    XMAS              = 9
+    TUTORIAL          = 10
+    MO                = 11
+    LP                = 12
+    POOL1             = 13
+    FH                = 14
+    CUSTOM            = 15
+    CD                = 16
+    BD                = 17
+    ABILITY_DRAFT     = 18
+    EVENT             = 19
+    ARDM              = 20
+    MID1V1            = 21
+    ALL_DRAFT         = 22
+    TURBO             = 23
+    MUTATION          = 24
+    COACHES_CHALLENGE = 25
```
These should be pascal case if they are more than 2 letters or not NONE (leave that as is)

Gobot1234 commented on steam/ext/dota2/state.py:
```diff
@@ -0,0 +1,29 @@
+    from .client import Client
+
+
+class GCState(GCState_[Any]):  # todo: implement basket-analogy for dota2
```
This needs to have support for _store_user overwriting the type of the User stored as you have a custom one

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,197 @@
+        self.sort_score: int = proto.sort_score
+        self.spectators: int = proto.spectators
+
+        self.start_time = datetime.datetime.fromtimestamp(proto.activate_time, datetime.timezone.utc)
+        self.game_time = datetime.timedelta(seconds=proto.game_time)
+        self.end_time = datetime.timedelta(seconds=proto.deactivate_time)
+        self.delay = datetime.timedelta(seconds=proto.delay)
+        self.last_update_time = datetime.datetime.fromtimestamp(proto.last_update_time, datetime.timezone.utc)
```
Please use steam.utils.DateTime.from_timestamp here and import timedelta directly from datetime

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,197 @@
+    def __init__(self, state: GCState, proto: CSourceTVGameSmall) -> None:
+        self._state: GCState = state
+
+        self.match_id: int = proto.match_id
+        self.server_steam_id: int = proto.match_id
+        self.lobby_id: int = proto.lobby_id
+
+        self.lobby_type = CSODOTALobbyLobbyType.try_value(proto.lobby_type)
+        self.game_mode = DOTAGameMode.try_value(proto.game_mode)
+        self.average_mmr: int = proto.average_mmr
+        self.sort_score: int = proto.sort_score
+        self.spectators: int = proto.spectators
```
The int annotations are unnecessary here
```suggestion
        self.match_id = proto.match_id
        self.server_steam_id = proto.match_id  # is this correct?
        self.lobby_id = proto.lobby_id

        self.lobby_type = CSODOTALobbyLobbyType.try_value(proto.lobby_type)
        self.game_mode = DOTAGameMode.try_value(proto.game_mode)
        self.average_mmr = proto.average_mmr
        self.sort_score = proto.sort_score
        self.spectators = proto.spectators
```

Aluerie replied:
Really? while browsing `discord.py` and etc source code, it looked like people like including those "reminder" type hints. Now I'm not sure when i'm supposed to type-hint in `__init__` 🤔 

Oups, `server_steam_id` is a silly error.

Gobot1234 replied:
It has them purely to make pyright happy in the type completeness score but I don't care, my code is typed and the type hints work for people. I'm not trying to up my score in a pretty pointless metric

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,197 @@
+    """
+
+    def __init__(self, state: GCState, proto: CSourceTVGameSmall) -> None:
+        self._state: GCState = state
```
```suggestion
        self._state = state
```

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,197 @@
+            self.tournament = TournamentMatch(
+                proto.league_id,
+                proto.series_id,
+                [
+                    TournamentTeam(proto.team_id_radiant, proto.team_name_radiant, proto.team_logo_radiant),
+                    TournamentTeam(proto.team_id_dire, proto.team_name_dire, proto.team_logo_dire),
+                ],
```
Should this be a tuple if it's fixed length?

Aluerie replied:
At first, thought about it but csgo `.teams` is typehinted as `: list[Team]` so I went with same... Isn't it also 2 teams in csgo? I'm not a player of it.

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,197 @@
+        # Since Immortal Draft update, players come from the proto message in a wrong order
+        # which can be fixed back with extra fields that they introduced later: `team`, `team_slot`
+        # why valve chose to introduce extra bytes fields instead of resorting it once after player selection - no clue
+        sorted_players = [p for p in sorted(proto.players, key=lambda player: (player.team, player.team_slot))]
```
```suggestion
        sorted_players = sorted(proto.players, key=attrgetter("team", "team_slot"))
```

Gobot1234 commented on steam/ext/dota2/client.py:
```diff
@@ -0,0 +1,176 @@
+
+        Parameters
+        ----------
+        hero_id
```
```suggestion
        hero
```

Gobot1234 commented on steam/ext/dota2/client.py:
```diff
@@ -0,0 +1,176 @@
+            Note, in this case Game Coordinator will still use only current top100 live matches, i.e. requesting
+            "filter by Muerta" will return only subset of those matches in which Muerta is currently being played.
+            It will not look into lower MMR games than top100. This behavior is consistent with how Watch Tab works.
+        max_matches
```
```suggestion
        limit
```
Everywhere else calls this a limit so for consistency sake please keep it the same

Gobot1234 commented on steam/ext/dota2/client.py:
```diff
@@ -0,0 +1,176 @@
+
+    async def top_live_matches(
+        self,
+        hero: Hero = MISSING,
+        max_matches: int = 100,
```
```suggestion
        hero: Hero = MISSING,
        max_matches: int = 100,
```
These should probably be kwarg only

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,197 @@
+    def __init__(self, state: GCState, proto: CSourceTVGameSmall) -> None:
+        self._state: GCState = state
+
+        self.match_id: int = proto.match_id
```
```suggestion
        self.id = proto.match_id
```
If this is a match, match.match_id reads badly to me

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,197 @@
+                battle_cup,
+            )
+
+        self.radiant_lead: int = proto.radiant_lead
+        self.radiant_score: int = proto.radiant_score
+        self.dire_score: int = proto.dire_score
+        self.building_state: float = proto.building_state  # todo: helper function to decode this into human-readable
+
+        self.custom_game_difficulty: int = proto.custom_game_difficulty
```
```suggestion
        self.radiant_lead = proto.radiant_lead
        self.radiant_score = proto.radiant_score
        self.dire_score = proto.dire_score
        self.building_state = proto.building_state  # todo: helper function to decode this into human-readable

        self.custom_game_difficulty = proto.custom_game_difficulty
```

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,197 @@
+
+class ProfileCard(Generic[UserT]):
+    def __init__(self, user: UserT, proto: CMsgDOTAProfileCard):
+        print(proto)
```
```suggestion
```
😉

Gobot1234 commented on steam/ext/dota2/protobufs/common.py:
```diff
@@ -0,0 +1,100 @@
+from ..enums import EMsg
+
+
+# fmt: off
+class CMsgDOTAProfileCardEStatID(betterproto.Enum):
+    Wins               = 3
+    Commends           = 4
+    GamesPlayed        = 5
+    FirstMatchDate     = 6
+    PreviousSeasonRank = 7
+    GamesMVP           = 8
+# fmt: on
```
betterproto enums don't need formatting like the regular ones because they shouldn't be exposed
```suggestion
class CMsgDOTAProfileCardEStatID(betterproto.Enum):
    Wins               = 3
    Commends           = 4
    GamesPlayed        = 5
    FirstMatchDate     = 6
    PreviousSeasonRank = 7
    GamesMVP           = 8
```

Gobot1234 commented on steam/ext/dota2/protobufs/sdk.py:
```diff
@@ -0,0 +1,65 @@
+from ..enums import EMsg
+
+
+# fmt: off
+class ESourceEngine(betterproto.Enum):
+    Source1 = 0
+    Source2 = 1
+
+
+class PartnerAccountType(betterproto.Enum):
+    NONE          = 0
+    PERFECT_WORLD = 1
+    INVALID       = 3
+# fmt: on
```
```suggestion
class ESourceEngine(betterproto.Enum):
    Source1 = 0
    Source2 = 1


class PartnerAccountType(betterproto.Enum):
    NONE          = 0
    PERFECT_WORLD = 1
    INVALID       = 3
```

Gobot1234 commented on steam/ext/dota2/protobufs/lobby.py:
```diff
@@ -0,0 +1,50 @@
+
+
+# fmt: off
+class CSODOTALobbyLobbyType(IntEnum):
```
```suggestion
class LobbyType(IntEnum):
```

Gobot1234 commented on steam/ext/dota2/protobufs/shared_enums.py:
```diff
@@ -0,0 +1,79 @@
+
+
+# fmt: off
+class DOTAGameMode(IntEnum):
```
```suggestion
class GameMode(IntEnum):
```

Gobot1234 reviewed (changes requested):
Overall looks very good, nearly there.

> Last time I had similar conversation (with Riot Games API wrapper developers) they actually had the opposite opinion - that it's better for both library devs and end-user to just follow "bad naming/practices/models" from the original devs instead of rethinking/over-cooking

I wouldn't disagree if the names were consistent at the very least but with steam they aren't

Aluerie committed:
🆘Apply suggestions from code review by Gobot

Co-authored-by: James Hilton-Balfe <gobot1234yt@gmail.com>

Aluerie committed:
🩺 More feedback corrections

Aluerie committed:
🪿Remove `fmt: off` from betterproto Enums

Aluerie committed:
🤺Quotes ' -> "

Black didn't automate it cause fmt: off 😒custom rule-set when

Aluerie committed:
🪥 Few more brush ups

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie committed:
🪥and more

Aluerie committed:
📺LiveMatchPlayer should subclass Partial user

Aluerie committed:
🦸‍♀️heroes property for LiveMatch

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie committed:
🤧fix import * abd `__all__`

Aluerie committed:
Merge branch 'introduce-ext.dota2' of https://github.com/Aluerie/steam.py into introduce-ext.dota2

Aluerie committed:
🔣Doc String + Remove C prefix

Aluerie committed:
📈 Glicko Rating and Behavior Summary

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie committed:
💥MatchDetails, MM stats, rating - very raw

Idk, is it too bad if I dont title stuff properly

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie committed:
Merge branch 'Gobot1234:main' into introduce-ext.dota2

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie committed:
Merge branch 'Gobot1234:main' into introduce-ext.dota2

Aluerie committed:
🏅Implement RankTier Medals

and a few brush ups

Aluerie committed:
🔬Introduce Match Minimal

Aluerie committed:
📂Divide `models.py` into a folder

Aluerie committed:
Create models.py

Aluerie committed:
🧓Match History

Aluerie committed:
📲Social Feed Post Message

Aluerie committed:
🪥Some Brush-ups

Erm

Aluerie committed:
〽️Instantiate Partial Match

Aluerie committed:
🖼️Profile Request

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie committed:
Doc Edits

Aluerie committed:
Merge branch 'introduce-ext.dota2' of https://github.com/Aluerie/steam.py into introduce-ext.dota2

Aluerie committed:
🔮Refactor things back to proper state -> models

Also got rid of the folder `models`; it was a bad idea.

Aluerie committed:
🔝Separate state for TopSource too

Aluerie committed:
🦜Separate state for MM stats

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie committed:
🌐 `replay_url`, `metadata_url`

Aluerie committed:
🙂Add Facets to protos

Protobufs were updated :x

Aluerie committed:
🎶Fix Match History

Aluerie committed:
🆕Add Kez hero

also fix for OD type 🤦‍♀️

Aluerie committed:
🪥`timeout` keyword arg for some state methods

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie committed:
Merge branch 'Gobot1234:main' into introduce-ext.dota2

Aluerie committed:
Bump typing-extensions from 4.10.0 to 4.12.2

Aluerie committed:
Revert "Bump typing-extensions from 4.10.0 to 4.12.2"

This reverts commit 186fe708016f0681182259a0dc475246a8af932a.

Aluerie committed:
Merge branch 'Gobot1234:main' into introduce-ext.dota2

Aluerie committed:
⚒️Rename `instantiate` to `create` methods

Also sneaky doc editing

Gobot1234 reviewed (commented):
Code looks pretty good. I would be hesitant to add a timeout parameter to general fetch methods as no where else in the library has this at least as a parameter bedside Client.wait_for

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,490 @@
+
+class ProfileCard:
+    def __init__(self, proto: common.ProfileCard):
+        self.account_id = proto.account_id
```
This should have a user attribute rather than account id which should be at least a partial user

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,490 @@
+        return f"<{self.__class__.__name__} id={self.id} hero={self.hero!r}>"
+
+
+class MatchMinimalPlayer:
```
This should subclass PartialUser

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,490 @@
+        self.team_number = proto.team_number
+
+
+class MatchMinimal:
```
Might be useful to have all of these subclass a common base that sets the basic attributes a match is expected to have that is a subclass of PartialMatch

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,490 @@
+        # which can be fixed back with extra fields that they introduced later: `team`, `team_slot`
+        sorted_players = sorted(proto.players, key=attrgetter("team", "team_slot"))
+
+        self.players: list[LiveMatchPlayer] = []
```
This should be a list comp rather than a for loop

Gobot1234 commented on steam/ext/dota2/models.py:
```diff
@@ -0,0 +1,490 @@
+        self._state = state
+
+        self.id = proto.match_id
+        self.server_steam_id = proto.server_steam_id
```
Can you add a method to fetch the full server and turn this into a steam.ID

Gobot1234 commented on steam/ext/dota2/state.py:
```diff
@@ -0,0 +1,226 @@
+    from ...types.id import ID32, ID64, Intable
+    from .client import Client
+
+    # TODO: test this whole Kwargs thing, does it error if we don't provide some of them?
```
Not required means the Client won't complain if they aren't passed to the initialiser. The defaults should be handled elsewhere

Gobot1234 commented on steam/ext/dota2/state.py:
```diff
@@ -0,0 +1,226 @@
+        futures = [
+            self.ws.gc_wait_for(
+                watch.GCToClientFindTopSourceTVGamesResponse,
+                check=partial(check, start_game),
```
This can be a regular old lambda function I think

Gobot1234 commented on steam/ext/dota2/state.py:
```diff
@@ -0,0 +1,226 @@
+        ]
+        await self.ws.send_gc_message(watch.ClientToGCFindTopSourceTVGames(**kwargs))
+        async with asyncio.timeout(timeout):
+            responses = await asyncio.gather(*futures)
```
```suggestion
            return await asyncio.gather(*futures)
```

Aluerie committed:
🪥Remove parameter `timeout`

Now all fetch methods are timeouted by `TIMEOUT_DEFAULT`

Aluerie committed:
🪥Names probably should be adjective+noun style

Aluerie committed:
🪥Use `lambda` instead of `partial`

Rip strict pylance/ruff complaints

Aluerie committed:
🆕Initial impl for `BaseMatch` and `BasePlayer`

Aluerie committed:
Update models.py

Gobot1234 commented on steam/ext/dota2/state.py:
```diff
@@ -3,7 +3,6 @@
 from __future__ import annotations
 
 import asyncio
-from functools import partial
```
I didn't realise that this was in a loop and using lambdas leads to a bug where the check is only getting the last start match id (I think). So this might need reverting

Aluerie committed:
🔙Revert `partial`->`lambda` change

Aluerie committed:
🪥Use `get_partial_user` over direct user constract

in `ProfileCard`

Aluerie committed:
Merge branch 'Gobot1234:main' into introduce-ext.dota2

Aluerie committed:
🪥 Add friends type hint

Aluerie committed:
Revert "🪥 Add friends type hint"

This reverts commit 8b7cbd83c29f601581eaa666e84c476c5f417057.

Aluerie committed:
🍫chore: Add Largo

Aluerie committed:
✨Add `create_from_npc_dota_hero_name` for `Hero`

Aluerie committed:
`zstd` memes ? it doesn't work for me otherwise

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Aluerie committed:
Merge branch 'Gobot1234:main' into introduce-ext.dota2
